module3/src/MyList.java

```diff
@@ -0,0 +1,697 @@
+        if (size + sizeToAdd >= capacity - GROW_FACTOR) {
+            capacity += GROW_FACTOR;
+            if(capacity < size + sizeToAdd) {
+                capacity = size + sizeToAdd + GROW_FACTOR;
+            }
+            Object[] result = new Object[capacity];
+            for (int i = 0; i < size; i++) {
+                result[i] = this.elementData[i];
+            }
+            this.elementData = result;
+        }
+
+        int i = 0;
+        for (Object o : c) {
+            this.elementData[size+i] = o;
+        }
+        size +=i;
+
+        return true;
+    }
+
+    /**
+     * Inserts all of the elements in the specified collection into this
+     * list at the specified position (optional operation).  Shifts the
+     * element currently at that position (if any) and any subsequent
+     * elements to the right (increases their indices).  The new elements
+     * will appear in this list in the order that they are returned by the
+     * specified collection's iterator.  The behavior of this operation is
+     * undefined if the specified collection is modified while the
+     * operation is in progress.  (Note that this will occur if the specified
+     * collection is this list, and it's nonempty.)
+     *
+     * @param index index at which to insert the first element from the
+     *              specified collection
+     * @param c collection containing elements to be added to this list
+     * @return <tt>true</tt> if this list changed as a result of the call
+     * @throws UnsupportedOperationException if the <tt>addAll</tt> operation
+     *         is not supported by this list
+     * @throws ClassCastException if the class of an element of the specified
+     *         collection prevents it from being added to this list
+     * @throws NullPointerException if the specified collection contains one
+     *         or more null elements and this list does not permit null
+     *         elements, or if the specified collection is null
+     * @throws IllegalArgumentException if some property of an element of the
+     *         specified collection prevents it from being added to this list
+     * @throws IndexOutOfBoundsException if the index is out of range
+     *         (<tt>index &lt; 0 || index &gt; size()</tt>)
+     */
+    @Override
+    public boolean addAll(int index, Collection<? extends T> c) {
+        throw new UnsupportedOperationException(); //return false; //Stanislav said we do not need to implement it
+    }
+
+    /**
+     * Removes all of the elements from this list (optional operation).
+     * The list will be empty after this call returns.
+     *
+     * @throws UnsupportedOperationException if the <tt>clear</tt> operation
+     *         is not supported by this list
+     */
+    @Override
+    public void clear() {
+        for (int i = 0; i < size; i++) {
+            this.elementData[i] = null;
+        }
+        this.size = 0;
+        this.capacity = DEFAULT_CAPACITY;
```

Comment:
you expect that capacity will be 10, but what if I do `clear` operation on your list which has 100500 elements? Exact capacity will be 100500, but your list will think that it has only 10

Reply:
TODO from here

Reply:
Added line:
`this.elementData = new Object[DEFAULT_CAPACITY];`

Reply:
But after this we cannot support operations like `list.subList(from, to).clear();`...
That's why I did not want to add this line previously...
I need to think how to resolve this dilemma